### PR TITLE
feat (api): startup and router

### PR DESCRIPTION
**Pull request by fsamin**
WIP

### CODE REVIEW

**Comment by yesnault** [comment]
on sha 6eb34745 :
```
➜  engine git:(ye-starter-hatchery) ✗ ./engine start api --config /Users/yesnault/src/github.com/ovh/cds/engine/dev.yml
Reading configuration file /Users/yesnault/src/github.com/ovh/cds/engine/dev.yml
While parsing config: yaml: line 2: did not find expected <document start>
➜  engine git:(ye-starter-hatchery) ✗ head -n 3 dev.yml
[API]

  [API.Artifact]
```

you update / modifiy some vendor ?

**Comment by fsamin** [comment]
humm try `go get -u github.com/pelletier/go-toml`

You create config with `./engine config new` ?

**Comment by yesnault** [comment]
go get -u both + govendor update both, then

```bash
$ go build
$ ./engine config new > dev.yml
$ ./engine start api --config dev.yml
Reading configuration file dev.yml
While parsing config: yaml: line 3: did not find expected <document start>
$ head -n 5 dev.yml

[API]

  [API.Artifact]
    Mode = "local"
```

**Comment by fsamin** [comment]
wait... yaml ???

try `toml` and viper will read toml :)

**Comment by yesnault** [comment]
🦆🦆🦆
```
./engine start api --config dev.toml
Reading configuration file dev.toml
Starting service api
```

File: engine/Makefile [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+TARGET_ARCH = amd64 arm
+
+GO_BUILD = go build
+GO_LINT = ${GOPATH}/bin/gometalinter
+GO_COV_MERGE = ${GOPATH}/bin/gocovmerge
+GO_GOVERALLS = ${GOPATH}/bin/goveralls
+GO_GO2XUNIT = ${GOPATH}/bin/go2xunit
+
+TEST_PKGS = $(shell go list ./... | grep -v vendor)
+TEST_DB_HOST = localhost
```
Comment (sguiheux) [inline engine/Makefile]: env var for DB values ?

File: engine/api/api.go [hunk 2]
```diff
@@ -152,29 +154,115 @@ type DefaultValues struct {
...
+		if err := os.MkdirAll(aConfig.Directories.Keys, os.FileMode(0600)); err != nil {
+			return fmt.Errorf("Unable to create directory %s: %v", aConfig.Directories.Keys, err)
+		}
+		log.Info("Directory %s has been created", aConfig.Directories.Keys)
+	} else if err != nil {
+		return fmt.Errorf("Invalid keys directory: %v", err)
+	}
+
+	switch aConfig.Artifact.Mode {
+	case "local", "openstack":
```
Comment (sguiheux) [inline engine/api/api.go]: swift?

File: engine/api/api.go [hunk 3]
```diff
@@ -393,45 +479,55 @@ func (a *API) Serve(ctx context.Context) error {
...
+		case <-ctx.Done():
+			log.Warning("Cleanup SQL connections")
+			s.Shutdown(ctx)
+			a.DBConnectionFactory.Close()
+			event.Publish(sdk.EventEngine{Message: "shutdown"})
+			event.Close()
+		}
+	}()
+
+	pipeline.Store = a.Cache
```
Comment (sguiheux) [inline engine/api/api.go]: twice?

File: engine/api/api_routes.go [hunk 4]
```diff
@@ -291,7 +299,7 @@ func (api *API) InitRouter() {
 	r.Handle("/workflow/hook/{model}", r.GET(api.getWorkflowHookModelHandler), r.POST(api.postWorkflowHookModelHandler, NeedAdmin(true)), r.PUT(api.putWorkflowHookModelHandler, NeedAdmin(true)))
 
 	// SSE
-	r.Handle("/mon/lastupdates/events", r.GET(lastUpdateBroker.ServeHTTP))
+	r.Handle("/mon/lastupdates/events", r.GET(api.LastUpdateBroker.ServeHTTP))
```
Comment (sguiheux) [inline engine/api/api_routes.go]: can be private ?

**Pull request closed without merging**